src/cluster.py

```diff
@@ -0,0 +1,212 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""The cluster module manages cluster-wide operations."""
+
+import json
+import logging
+import subprocess
+import uuid
+from typing import Tuple
+
+import ops.charm
+import ops_sunbeam.guard as sunbeam_guard
+import tenacity
+from charms.operator_libs_linux.v2 import snap
+
+import charm
+import microceph
+import relation_handlers
+from microceph import CephHealth, CephStatus
+
+logger = logging.getLogger(__name__)
+
+
+class ClusterNodes(ops.framework.Object):
+    """ClusterNodes manages adding and joining nodes to the microceph cluster."""
+
+    def __init__(self, charm: "charm.MicroCephCharm"):
+        super().__init__(charm, "cluster-nodes")
+        self.charm = charm
+
+    def add_node_to_cluster(self, event: ops.framework.EventBase) -> None:
+        """Add node to microceph cluster."""
+        if not event.unit:
+            return
+        cmd = ["sudo", "microceph", "cluster", "add", event.unit.name]
+        try:
+            logger.debug(f'Running command {" ".join(cmd)}')
+            process = subprocess.run(cmd, capture_output=True, text=True, check=True, timeout=180)
+            logger.debug(f"Command finished. stdout={process.stdout}, " f"stderr={process.stderr}")
+            token = process.stdout.strip()
+            self.charm.peers.set_app_data({f"{event.unit.name}.join_token": token})
+        except (subprocess.CalledProcessError, subprocess.TimeoutExpired) as e:
+            logger.warning(e.stderr)
+            error_node_already_exists = (
+                'Failed to create "internal_token_records" entry: UNIQUE '
+                "constraint failed: internal_token_records.name"
+            )
+            if error_node_already_exists not in e.stderr:
+                raise e
+
+    def join_node_to_cluster(self, event: ops.framework.EventBase) -> None:
+        """Join node to microceph cluster."""
+        if not event.unit:
+            return
+        token = self.charm.peers.get_app_data(f"{event.unit.name}.join_token")
+        cmd = ["sudo", "microceph", "cluster", "join", token]
+        try:
+            logger.debug(f'Running command {" ".join(cmd)}')
+            process = subprocess.run(cmd, capture_output=True, text=True, check=True, timeout=180)
+            logger.debug(f"Command finished. stdout={process.stdout}, " f"stderr={process.stderr}")
+            self.charm.peers.interface.state.joined = True
+            self.charm.peers.set_unit_data({"joined": json.dumps(True)})
+        except (subprocess.CalledProcessError, subprocess.TimeoutExpired) as e:
+            logger.warning(e.stderr)
+            raise e
+
+
+class ClusterUpgrades(ops.framework.Object):
+    """ClusterUpgrades manages snap upgrades across the cluster."""
+
+    charm = None
+    _stored = ops.framework.StoredState()
+
+    def __init__(self, charm: "charm.MicroCephCharm"):
```

Comment:
Is `"charm.MicroCephCharm"` a special value or we have a `string` default value for a charm object ?

Reply:
No, this is a type hint -- an example of specifying a type via a forward reference so you don't have to import the actual type here

Reply:
Ack